R/moma_sfcca.R

```diff
@@ -0,0 +1,903 @@
+                            xlab = "X features"
+                        )
+                        axis(1,
+                            at = 1:self$px,
+                            labels = self$x_coln
+                        )
+                    })
+
+                    output$lambda_y <- renderPrint({
+                        k <- as.integer(input$rank)
+
+                        alpha_x_value <- get_rank_k_result()$chosen_alpha_x[k]
+                        cat(paste0("alpha_x = ", alpha_x_value), "\n")
+                        alpha_y_value <- get_rank_k_result()$chosen_alpha_y[k]
+                        cat(paste0("alpha_y = ", alpha_y_value), "\n")
+                        lambda_x_value <- get_rank_k_result()$chosen_lambda_x[k]
+                        cat(paste0("lambda_x = ", lambda_x_value), "\n")
+                        lambda_y_value <- get_rank_k_result()$chosen_lambda_y[k]
+                        cat(paste0("lambda_y = ", lambda_y_value), "\n")
+                    })
+
+                    output$X_rows_projected <- renderPlot({
+                        X_projected <- get_X_projected()$proj_data
+
+                        # print(X_projected)
+                        plot(X_projected,
+                            xlab = "PC1",
+                            ylab = "PC2",
+                            main = "Projected rows of the data matrix X"
+                        )
+                    })
+
+                    output$Y_rows_projected <- renderPlot({
+                        Y_projected <- get_Y_projected()$proj_data
+
+                        # print(X_projected)
+                        plot(Y_projected,
+                            xlab = "PC1",
+                            ylab = "PC2",
+                            main = "Projected rows of the data matrix Y"
+                        )
+                    })
+                }
+            )
+        }
+    )
+)
+
+#' The Deflation Scheme for CCA
+#'
+#' In \code{MoMA} one deflation scheme is provided for CCA.
+#'
+#' Let \eqn{X,Y} be two data matrices (properly scaled and centered) of the same number of
+#' rows. Each row represents a sample. The penalized CCA problem is formulated as
+#'
+#' \eqn{ \min_{u,v} \, u^T X^T Y v + \lambda_u P_u(u) + \lambda_v P_v(v)  }
+#'
+#' \eqn{ \text{s.t. } \| u \|_{I+\alpha_u \Omega_u} \leq 1, \| v \|_{I + \alpha_v \Omega_v} \leq 1.  }
+#'
+#' In the discussion below, let \eqn{u,v} be the solution to the above problem.
+#' Let \eqn{c_x = Xu, c_y = Yv}. The deflation scheme is as follow:
+#'
+#' \eqn{X \leftarrow  { X } -  { c_x } \left(  { c_x } ^ { T }  { c_x } \right) ^ { - 1 }  { c_x } ^ { T }  { X }
+#' = ( I - { c_x } \left(  { c_x } ^ { T }  { c_x } \right) ^ { - 1 }  { c_x } ^ { T } )X,}
+#'
+#' \eqn{ Y \leftarrow { Y } -  { c_y } \left(  { c_y } ^ { T }  { c_y } \right) ^ { - 1 }  { c_y } ^ { T }  { Y }
+#' = (I -  { c_y } \left(  { c_y } ^ { T }  { c_y } \right) ^ { - 1 }  { c_y } ^ { T } ) Y}.
```

Comment:
Isn't this just "normal" (PCA-style) projection deflation on the X^TY matrix? That might an easier way to implement. I'm not sure we need to break these up at the C++ level.

Reply:
It might be useful to bookkeep the deflated X and Y.